app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/TrackBackgrounds.kt

```diff
@@ -0,0 +1,96 @@
+
+import android.content.Context
+import android.text.TextUtils
+import org.xmlpull.v1.XmlPullParser
+import java.util.HashMap
+import java.util.LinkedHashMap
+import nerd.tuxmobil.fahrplan.congress.R
+
+object TrackBackgrounds {
+    fun getHashMapResource(c: Context, hashMapResId: Int): Map<String?, String?>? {
+        var map: MutableMap<String?, String?>? = null
+        val parser = c.resources.getXml(hashMapResId)
+        var key: String? = null
+        var value: String? = null
+        try {
+            var eventType = parser.eventType
+            while (eventType != XmlPullParser.END_DOCUMENT) {
+                when (eventType) {
+//                  XmlPullParser can also be START_DOCUMENT
+                    XmlPullParser.START_TAG -> {
+                        if (parser.name == "map") {
+                            val isLinked = parser.getAttributeBooleanValue(null, "linked", false)
+                            map = if (isLinked) LinkedHashMap() else HashMap()
+                        } else if (parser.name == "entry") {
+                            key = parser.getAttributeValue(null, "key")
+                            if (null == key) {
+                                parser.close()
+
+                                return null
+                            }
+                        }
+                    }
+                    XmlPullParser.END_TAG -> {
+                        if (parser.name == "entry") {
+                            map!![key] = value
+                            key = null
+                            value = null
+                        }
+                    }
+                    XmlPullParser.TEXT -> {
+                        if (key != null) {
+                            value = parser.text.trim { it <= ' ' }
+                        }
+                    }
+                }
+                eventType = parser.next()
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+            return null
+        }
+        return map
+    }
+
+    fun buildTrackBackgroundHashMap(
+        trackNamesMap: Map<String?, String?>?,
+        prefix: String?,
+        resourceType: String?,
+        context: Context
+    ): Map<String?, Int> {
+        val trackNameResourceIdPairs = HashMap<String?, Int>()
+        for ((key, value1) in trackNamesMap!!) {
+            var value = prefix
+            // Handle empty track names
+            // key can have the value: ""
+            // See track_resource_names.xml
+            if (!TextUtils.isEmpty(key)) {
+                value += "_$value1"
+            }
+            val resourceId = context.resources.getIdentifier(value, resourceType, context.packageName)
+            trackNameResourceIdPairs[key] = resourceId
+        }
+        return trackNameResourceIdPairs
+    }
+
+    fun getTrackNameBackgroundColorDefaultPairs(context: Context) = buildTrackBackgroundHashMap(
+        getHashMapResource(
```

Comment:
Given that `buildTrackBackgroundHashMap()` cannot deal with null being returned from `getHashMapResource()` I would rather put `!!` behind `getHashMapResource()` here and below to **fail fast** and not within the called function. The parameter can then become `trackNamesMap: Map<String?, String?>` instead of `trackNamesMap: Map<String?, String?>?`.